pandas/core/algorithms.py

```diff
@@ -614,6 +618,15 @@ def factorize(
     values = _ensure_arraylike(values)
     original = values
 
+    if isinstance(
+        values,
+        (ABCDatetimeIndex, ABCTimedeltaIndex, ABCDatetimeArray, ABCTimedeltaArray),
+    ):
+        # Defer to the method in order to retain freq
+        if sort is False and size_hint is None:
```

Comment:
Is the check for `sort is False` needed? The sorting or not is handled below (if you it to `codes, uniques = ..` instead of directly returning)